linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala

```diff
@@ -2016,29 +2024,37 @@ private[emitter] class FunctionEmitter(sjsGen: SJSGen) {
             redo(CreateJSClass(className, newCaptureValues))(env)
           }
 
-        case _ =>
-          if (lhs == Lhs.Discard) {
-            /* Go "back" to transformStat() after having dived into
-             * expression statements. Remember that Lhs.Discard is a trick that
-             * we use to "add" all the code of pushLhsInto() to transformStat().
-             */
-            rhs match {
-              case _:Skip | _:VarDef | _:Assign | _:While |
-                  _:Debugger | _:JSSuperConstructorCall | _:JSDelete |
-                  _:StoreModule | Transient(_:SystemArrayCopy) =>
-                transformStat(rhs, tailPosLabels)
-              case _ =>
-                throw new IllegalArgumentException(
-                    "Illegal tree in JSDesugar.pushLhsInto():\n" +
-                    "lhs = " + lhs + "\n" +
-                    "rhs = " + rhs + " of class " + rhs.getClass)
-            }
-          } else {
-            throw new IllegalArgumentException(
-                "Illegal tree in JSDesugar.pushLhsInto():\n" +
-                "lhs = " + lhs + "\n" +
-                "rhs = " + rhs + " of class " + rhs.getClass)
+        // Statement-only trees
+
+        case _:Skip | _:VarDef | _:Assign | _:While | _:Debugger |
+            _:JSSuperConstructorCall | _:JSDelete | _:StoreModule |
+            Transient(_:SystemArrayCopy) =>
+          /* Go "back" to transformStat() after having dived into
+           * expression statements. This can only happen for Lhs.Discard and
+           * for Lhs.Return's whose target is a statement.
+           */
+          lhs match {
+            case Lhs.Discard =>
+              transformStat(rhs, tailPosLabels)
+            case Lhs.ReturnFromFunction =>
+              js.Block(
+                  transformStat(rhs, tailPosLabels = Set.empty),
+                  js.Return(js.Undefined()))
```

Comment:
It feels odd that we do not have tail label optimization here, but IIUC that's on par with the status quo.

Reply:
We *can't* have tail label optimization here. Indeed, there *is* control-flow-disrupting code after the transformed `rhs`, namely this `js.Return(js.Undefined())`. The labels that are in tail position *after* the `js.Return(js.Undefined())` are definitely *not* in tail position of the `rhs`.

Reply:
Ah, sorry, I wasn't clear. I understand why `tailPosLabels = Set.empty` is necessary.

What I didn't get is why there is no attempt to eliminate the return we are emitting itself here.

But I think now I understand: We only get here if an actual label is transformed to an `Lhs.ReturnFromFunction`. If the function we are translating has void return type (and no top-level label), we invoke `transformStat` immediately.

Maybe that warrants a comment :P